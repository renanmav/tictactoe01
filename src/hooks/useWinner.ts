import { Dispatch, SetStateAction, useEffect, useState } from 'react';

/**
 *  Winner is:
 *    '?' if game is not finished
 *    'X' if game has been won by X
 *    'O' if game has been won by O
 *    'C' if game has been won by 'cat' (no moves left, but no winner)
 */
export type Winner = '?' | 'X' | 'O' | 'C';

/**
 * checkStalemate
 *  Checks if all cells in 'board' are occupied, and if they are,
 *  sets winner to 'C' (for 'cat')
 */
interface CheckStalemateArgs {
  board: string;
  setWinner: Dispatch<SetStateAction<Winner>>;
  setWinnerMatch: Dispatch<SetStateAction<number[]>>;
}
export function checkStalemate(args: CheckStalemateArgs) {
  const { board, setWinner, setWinnerMatch } = args;

  for (let i = 0; i < 9; i++) {
    if (board.charAt(i) === ' ') {
      return;
    }
  }

  setWinner('C');
  setWinnerMatch([]);
}

/**
 * checkVictory
 *  Checks if the indicated cells ('p1', 'p2' and 'p3') contain a
 *  three-in-a-row win for either player. If they do, winner will be
 *  set to 'X' or 'O'. It merely checks whether all three cells are
 *  occupied by 'X' or all three by 'O'.
 */
interface CheckVictoryArgs {
  board: string;
  setWinner: Dispatch<SetStateAction<Winner>>;
  setWinnerMatch: Dispatch<SetStateAction<number[]>>;
  p1: number;
  p2: number;
  p3: number;
}
export function checkVictory(args: CheckVictoryArgs) {
  const { board, setWinner, setWinnerMatch, p1, p2, p3 } = args;

  const c1 = board.charAt(p1);
  if (c1 === ' ') {
    return false;
  }

  const c2 = board.charAt(p2);
  if (c1 !== c2) {
    return false;
  }

  const c3 = board.charAt(p3);
  if (c1 !== c3) {
    return false;
  }

  setWinner(c1 as Winner);
  setWinnerMatch([p1, p2, p3]);
}

/**
 * useWinner
 *  Checks the board to see if the game is finished. If the game
 *  is done, winner is set to 'X', 'O' or 'C' according to whether
 *  the game was won by player X, player O or 'cat'. If the game
 *  is not finished, winner is set to '?'
 */
export function useWinner(board: string) {
  const [winner, setWinner] = useState<Winner>('?');
  const [isEvaluating, setIsEvaluating] = useState(false);
  const [winnerMatch, setWinnerMatch] = useState<number[]>([]);

  useEffect(() => {
    if (board === '         ') {
      return;
    }

    setIsEvaluating(true);

    checkVictory({ board, setWinner, setWinnerMatch, p1: 0, p2: 1, p3: 2 }) || // check for 3-in-a-row horizontally
    checkVictory({ board, setWinner, setWinnerMatch, p1: 3, p2: 4, p3: 5 }) ||
    checkVictory({ board, setWinner, setWinnerMatch, p1: 6, p2: 7, p3: 8 }) ||
    checkVictory({ board, setWinner, setWinnerMatch, p1: 0, p2: 3, p3: 6 }) || // check for 3-in-a-row vertically
    checkVictory({ board, setWinner, setWinnerMatch, p1: 1, p2: 4, p3: 7 }) ||
    checkVictory({ board, setWinner, setWinnerMatch, p1: 2, p2: 5, p3: 8 }) ||
    checkVictory({ board, setWinner, setWinnerMatch, p1: 0, p2: 4, p3: 8 }) || // check for 3-in-a-row diagonally
      checkVictory({ board, setWinner, setWinnerMatch, p1: 6, p2: 4, p3: 2 }) ||
      checkStalemate({ board, setWinner, setWinnerMatch });

    setIsEvaluating(false);

    return () => {};
  }, [board]);

  function reset() {
    setWinnerMatch([]);
    setWinner('?');
  }

  return { winner, isEvaluating, winnerMatch, reset };
}
