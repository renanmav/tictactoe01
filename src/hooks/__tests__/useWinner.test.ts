import { renderHook } from '@testing-library/react-hooks';

import { useWinner, checkStalemate, checkVictory } from '../useWinner';

describe('checkStalemate', () => {
  test('if winner is set to "C" when no moves are left', () => {
    const board = 'XXOOOXOXO';
    const setWinner = jest.fn();
    const setWinnerMatch = jest.fn();

    checkStalemate({ board, setWinner, setWinnerMatch });

    expect(setWinner).toBeCalledTimes(1);
    expect(setWinner).toBeCalledWith('C');
  });

  test('if winner is not set to "C" when moves are left', () => {
    const board = 'XXOOO OXO';
    const setWinner = jest.fn();
    const setWinnerMatch = jest.fn();

    checkStalemate({ board, setWinner, setWinnerMatch });

    expect(setWinner).not.toBeCalled();
  });
});

describe('checkVictory', () => {
  test('if "X" is set to winner when a victory happens on a row', () => {
    const board = '   XXX   ';
    const setWinner = jest.fn();
    const setWinnerMatch = jest.fn();

    checkVictory({ board, setWinner, setWinnerMatch, p1: 3, p2: 4, p3: 5 });

    expect(setWinner).toBeCalledTimes(1);
    expect(setWinner).toBeCalledWith('X');
    expect(setWinnerMatch).toBeCalledTimes(1);
    expect(setWinnerMatch).toBeCalledWith([3, 4, 5]);
  });

  test('if "O" is set to winner when a victory happens on a diagonal', () => {
    const board = 'OXXXOXXXO';
    const setWinner = jest.fn();
    const setWinnerMatch = jest.fn();

    checkVictory({ board, setWinner, setWinnerMatch, p1: 0, p2: 4, p3: 8 });

    expect(setWinner).toBeCalledTimes(1);
    expect(setWinner).toBeCalledWith('O');
    expect(setWinnerMatch).toBeCalledWith([0, 4, 8]);
  });
});

describe('useWinner', () => {
  test('if "?" is returned when the game is not finished', () => {
    const board = 'X        ';

    const { result } = renderHook(() => useWinner(board));

    expect(result.current.winner).toBe('?');
    expect(result.current.winnerMatch).toMatchSnapshot();
  });

  test('if "X" is returned when player X wins with 3-in-a-row horizontally', () => {
    const board = '      XXX';

    const { result } = renderHook(() => useWinner(board));

    expect(result.current.winner).toBe('X');
    expect(result.current.winnerMatch).toMatchSnapshot();
  });

  test('if "O" is returned when player O wins with 3-in-a-row vertically', () => {
    const board = 'OXXOXOO  ';

    const { result } = renderHook(() => useWinner(board));

    expect(result.current.winner).toBe('O');
    expect(result.current.winnerMatch).toMatchSnapshot();
  });

  test('if "C" is returned when a "cat" happens', () => {
    const board = 'XOXXXOOOX';

    const { result } = renderHook(() => useWinner(board));

    expect(result.current.winner).toBe('C');
    expect(result.current.winnerMatch).toMatchSnapshot();
  });
});
