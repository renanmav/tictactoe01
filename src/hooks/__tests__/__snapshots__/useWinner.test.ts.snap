// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useWinner if "?" is returned when the game is not finished 1`] = `Array []`;

exports[`useWinner if "C" is returned when a "cat" happens 1`] = `Array []`;

exports[`useWinner if "O" is returned when player O wins with 3-in-a-row vertically 1`] = `
Array [
  0,
  3,
  6,
]
`;

exports[`useWinner if "X" is returned when player X wins with 3-in-a-row horizontally 1`] = `
Array [
  6,
  7,
  8,
]
`;
